From 1762489ee9e0931b6a699cd6fdb92bfed3484cc1 Mon Sep 17 00:00:00 2001
From: "Ooi, Joyce" <joyce.ooi@intel.com>
Date: Mon, 16 Mar 2020 12:35:11 +0800
Subject: [PATCH 131/151] net: ethernet: altera: fix TOD adjust time

commit  2656b754b31c5e16231b5f8930fb385096ddd98c from
https://github.com/altera-opensource/linux-socfpga.git
branch is socfpga-5.4.64-lts

This patch fixes the TOD adjust time by adding a check on the delta
value before performing the coarse time offset adjustment. This prevents
a negative delta value to cause the master offset to be calculated
incorrectly.

Signed-off-by: Ooi, Joyce <joyce.ooi@intel.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
Integrated-by: Jun Zhang <jun.zhang@windriver.com>
---
 drivers/net/ethernet/altera/intel_fpga_tod.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/net/ethernet/altera/intel_fpga_tod.c b/drivers/net/ethernet/altera/intel_fpga_tod.c
index 0d381818b3c7..b0a800791123 100644
--- a/drivers/net/ethernet/altera/intel_fpga_tod.c
+++ b/drivers/net/ethernet/altera/intel_fpga_tod.c
@@ -190,6 +190,9 @@ static int intel_fpga_tod_adjust_time(struct ptp_clock_info *ptp, s64 delta)
 	 * just set the time.
 	 */
 	if (count > TOD_ADJUST_COUNT_MAX) {
+		if (neg_adj)
+			delta = -delta;
+
 		/* Perform the coarse time offset adjustment */
 		ret = coarse_adjust_tod_clock(priv, delta);
 	} else {
-- 
2.26.1

